File: src/UnwindMC/Generation/Ast/Transformer.fs
module Transformer

open Ast

type T = {
    transformExpression: T -> Expression -> Expression
    transformStatement: T -> Statement -> Statement
}

let private transformExpression (t: T) (expr: Expression): Expression =
    match expr with
    | Binary (op, left, right) -> Binary (op, t.transformExpression t left, t.transformExpression t right)
    | Dereference expr -> Dereference (t.transformExpression t expr)
    | Unary (op, operand) -> Unary (op, t.transformExpression t operand)
    | Value value -> Value value
    | VarRef var -> VarRef var

let private transformStatement (t: T) (statement: Statement): Statement =
    match statement with
    | Assignment (var, expr) -> Assignment (var, t.transformExpression t expr)
    | Break -> Break
    | Continue -> Continue
    | DoWhile (statements, condition) -> DoWhile (statements |> ROL.map (t.transformStatement t), t.transformExpression t condition)
    | For (condition, modifier, body) -> For (t.transformExpression t condition, modifier |> ROL.map (t.transformStatement t), body |> ROL.map (t.transformStatement t))
    | FunctionCall expr -> FunctionCall (t.transformExpression t expr)
    | IfThenElse (condition, trueBranch, falseBranch) ->
        IfThenElse (t.transformExpression t condition, trueBranch |> ROL.map (t.transformStatement t), falseBranch |> ROL.map (t.transformStatement t))
    | Return var -> Return var
    | While (condition, statements) -> While (t.transformExpression t condition, statements |> ROL.map (t.transformStatement t))

let transform t = t.transformStatement t

let def: T = {
    transformExpression = transformExpression
    transformStatement = transformStatement
}

File: src/transformer.rs
use ast::*;

pub trait Transformer {
    fn transform_statement(&mut self, stmt: Statement) -> Statement;
    fn transform_expression(&mut self, expr: Expression) -> Expression;
}

pub fn default_transform_statement(transformer: &mut Transformer, stmt: Statement) -> Statement {
    use ast::Statement::*;
    match stmt {
        Assignment(var, expr) => Assignment(var, transformer.transform_expression(expr)),
        Break => Break,
        Continue => Continue,
        DoWhile(body, condition) => DoWhile(
            body.into_iter()
                .map(|s| transformer.transform_statement(s))
                .collect(),
            transformer.transform_expression(condition),
        ),
        For(condition, modifier, body) => For(
            transformer.transform_expression(condition),
            modifier
                .into_iter()
                .map(|s| transformer.transform_statement(s))
                .collect(),
            body.into_iter()
                .map(|s| transformer.transform_statement(s))
                .collect(),
        ),
        FunctionCall(expr) => FunctionCall(transformer.transform_expression(expr)),
        IfThenElse(condition, true_branch, false_branch) => IfThenElse(
            transformer.transform_expression(condition),
            true_branch
                .into_iter()
                .map(|s| transformer.transform_statement(s))
                .collect(),
            false_branch
                .into_iter()
                .map(|s| transformer.transform_statement(s))
                .collect(),
        ),
        Return(var) => Return(var),
        While(condition, body) => While(
            transformer.transform_expression(condition),
            body.into_iter()
                .map(|s| transformer.transform_statement(s))
                .collect(),
        ),
    }
}

pub fn default_transform_expression(transformer: &mut Transformer, expr: Expression) -> Expression {
    use ast::Expression::*;
    match expr {
        Binary(op, left, right) => Binary(
            op,
            Box::new(transformer.transform_expression(*left)),
            Box::new(transformer.transform_expression(*right)),
        ),
        Dereference(expr) => Dereference(Box::new(transformer.transform_expression(*expr))),
        Unary(op, operand) => Unary(op, Box::new(transformer.transform_expression(*operand))),
        Value(value) => Value(value),
        VarRef(var) => VarRef(var),
    }
}
